OpenID Authentication

bsu3338 opened the pull request:
This code introduces the ability to configure OpenID as an authentication method similar to the current Google option.  
The method uses passport openid-client strategy which seems to be the most popular option. I looked at several passport openid stratgies and they all seem to require session support.  OpenID Connect requires session support for its operation, specifically for managing state and nonce parameters, which are used for CSRF attack prevention and replay attack prevention, respectively. Therefore, simply disabling session management would not be appropriate for OpenID Connect.

I attempted a hybrid approach that can give us the benefits of both JWT and sessions. You have the stateful session for the OpenID Connect to function correctly, and a stateless JWT for authorizing your users across services and requests, allowing better scalability.

It's a compromise that gives you the statelessness and scalability of JWT, but also the features of OpenID Connect that require session support. The JWT is still useful because it allows your clients to make authenticated requests without constantly going through the OpenID Connect flow, and it can be used with services that don't support OpenID Connect.

In conclusion, the session is needed for the OpenID Connect flow and the JWT is used for maintaining user authentication in a stateless manner after the OpenID Connect flow is completed.

Honestly OpenID and JWT is stretching my brain a bit.

Microsoft requires an access token to download the picture of the user through their graph api. So I made a choice of similar to the Stable Diffusion plugin to store the image in /images/openid and putting the link to the image in the avatar attribute.

I tried to implement many of the suggestions from my last pull request, shorter variable names, integration with JWT, picture support, checks in case OpenID does not return an email or picture link, checks if label or image is not provided in variables.

Added image as a volume so they are persistent across container restarts
 
~CAVEAT:
For some reason, I cannot get the OPEN_LABEL to change the button name. I thought it was working on my previous pull request.~
thanks @ClaraLeigh for helping me see my mistake
 
Again, thanks for all your help and suggestions

The documentation is within the .env.example


## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation update  
  

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration:
##
Tested using Azure AD

Azure AD
- Sign in to the Azure Portal
- Search for and select Azure Active Directory
- Select App registrations > New registration
- Enter a display Name
- Choose Web
- Enter https://fqdn/oauth/openid/callback
- Select Overview
- Note of the Application (client) ID and Directory (tenant) ID
- Select Authentication
- Check Access tokens and ID tokens
- Select Certificates & Secrets
- Create a secret and note the value

Edit .env to include variables
OPENID_CLIENT_ID=From Note Above
OPENID_CLIENT_SECRET=From Note Above
OPENID_ISSUER=https://login.microsoftonline.com/(tenant id)/v2.0/
OPENID_SESSION_SECRET=Random String
OPENID_SCOPE=openid profile email
OPENID_CALLBACK_URL=/oauth/openid/callback

### **Test Configuration**:
##


## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules


![image](https://github.com/danny-avila/LibreChat/assets/11950213/c4fc6aa8-98e3-4828-9520-6d73d492c6b7)

![image](https://github.com/danny-avila/LibreChat/assets/11950213/c45e1188-c562-48cf-a3d9-d87c421985c8)

bsu3338 committed:
Squashed commit of the following:

commit 26ab03fb36fcc7fcee63fdf3ae8c2dfb29027eff
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Tue Jun 13 00:23:23 2023 -0500

    Update Registration.spec.tsx

commit e908dd82fe9ef1b43c75ee64c183d2f654bdac1c
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Tue Jun 13 00:23:01 2023 -0500

    Update Login.spec.tsx

commit 223734820fb77d7fb5af4802af642d1c1fd7c1f5
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Tue Jun 13 00:22:39 2023 -0500

    Update Registration.tsx

commit 7036d3dd0538979ee397d958ebc113bb0ea32411
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Tue Jun 13 00:21:55 2023 -0500

    Update Login.tsx

commit 76bb78221db3195fd930fe9cfd6a5da7194fa759
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Tue Jun 13 00:21:03 2023 -0500

    Update envConstants.js

commit ee2f69f33d75fbb57022afbcd9564bca38a46bee
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Tue Jun 13 00:20:08 2023 -0500

    Update docker-compose.yml

commit 5ac72d789b3446884c6e2f4f595cbf67d731d43c
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Tue Jun 13 00:18:41 2023 -0500

    Update Dockerfile

commit d24341db2bd5b17eb89ab01e171a5f51f3beab0a
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Tue Jun 13 00:16:38 2023 -0500

    Update .env.example

commit 22154f4a09c5fcdfee95d43609fb01a5a883b7a9
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Tue Jun 13 00:07:48 2023 -0500

    Update Registration.spec.tsx

commit 5163f7d372a6a03c94f4357b358211a03369456e
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Tue Jun 13 00:07:30 2023 -0500

    Update Login.spec.tsx

commit 61da49e330a9376e130b24dc944854f97ab58d80
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Tue Jun 13 00:07:00 2023 -0500

    Update Registration.tsx

commit 0e45d3f0dbde34388ff2f0b2dc51b983b472eb05
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Tue Jun 13 00:06:18 2023 -0500

    Update Login.tsx

commit dca1e5367e5f3b468c7964218cc5914ca53095af
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Tue Jun 13 00:05:07 2023 -0500

    Update envConstants.js

commit f48c058465d82b03716ba85224e9f97007e014d2
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Tue Jun 13 00:04:05 2023 -0500

    Update .env.example

commit 818226c9cb079acae4fcbfe5997e4aa9e3c6d2cc
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:59:08 2023 -0500

    Update .env.example

commit 9a805439189b352a38ac7654d7a31bb28f0f58dd
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:58:31 2023 -0500

    Update env.d.ts

commit 3f37ce54758b017c9281b7fad9b040a47630ec66
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:57:04 2023 -0500

    Update .env.example

commit 1026036f4dd529e9531c53084450ce768cfca4c1
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:50:36 2023 -0500

    Update docker-compose.yml

commit a61cf7b8c51d4a9bd73a20bd67abc29891c11463
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:50:00 2023 -0500

    Update Dockerfile

commit 79610d6648755cd5ec45215b9fdbe04ba8242fcf
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:35:34 2023 -0500

    Update package-lock.json

commit e40853fd2b77f2db5be1c3dfd8b170d650e23271
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:30:17 2023 -0500

    Update envConstants.js

commit 5529bc61b43f279fb4418c3851be2f9011b6454d
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:25:58 2023 -0500

    Update docker-compose.yml

commit 07848cc464a64f7cad484e24a1310dc61aa03b18
Merge: ec628a3 72e9828
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:24:03 2023 -0500

    Merge branch 'danny-avila:main' into openid-client

commit ec628a3044ba963b4e733c72229400074e7c2bc4
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:23:16 2023 -0500

    Update envConstants.js

commit 21272221db0f58c244f08335482d45b177d338ab
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:21:59 2023 -0500

    Update Registration.spec.tsx

commit d3f2949c0484d5760e7b689501852f86209992a3
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:21:12 2023 -0500

    Update Login.spec.tsx

commit f2cf23ddd6708a3bb8d032dde5f1ce300dbe8cad
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:20:15 2023 -0500

    Update Registration.tsx

commit 482c346b2a7baf958665c9474223d2557504dee5
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:17:53 2023 -0500

    Update Login.tsx

commit 2f017aa5bf4ef91b73fe027fb346132e1a5d8b87
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:14:17 2023 -0500

    Update env.d.ts

commit addfd95cf93ef19cae05bab652d634af64313e6a
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:13:16 2023 -0500

    Create openidStrategy.js

commit 84c3b5c2f078494d8380f3a02e3ba2d935d8d79f
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:09:02 2023 -0500

    Update oauth.js

commit 63225cdf33b7f42005b4a446797acbd91b7ee4a7
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:07:35 2023 -0500

    Update index.js

commit 6efe4dafd4359ed1c3139468bf9d43f70bbaf6aa
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:04:55 2023 -0500

    Update package.json

commit 201badbbb5a5c8d48f5c4cba3a1349d4cfc7a070
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:03:37 2023 -0500

    Update User.js

commit 7d13d5c303465be9b1268e5f6d9bdf7bb8dfb2e4
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:02:29 2023 -0500

    Update Dockerfile

commit 2ef7f84ea77f281c3dce61211d9fd841a6424e65
Author: bsu3338 <bsu3338@users.noreply.github.com>
Date:   Mon Jun 12 23:00:42 2023 -0500

    Update .env.example

danny-avila commented:
Thanks for working on this!

bsu3338 committed:
Update openidStrategy.js

danny-avila commented:
for OPENID_LABEL, is this a frontend environment variable? you mentioned you couldn't get this showing.

This is related to #511, and still deliberating what route to take with the solution.

mitch-b commented:
I ran your branch and got an Internal Server Error on the OAuth callback. Has anyone else been able to successfully test the integration? 

I'll see if I can dig into the root cause. Hopefully user error, but wanted to keep the discussion going on this PR.

ClaraLeigh commented on .env.example:
```diff
@@ -202,5 +215,13 @@ VITE_APP_TITLE="LibreChat"
+# and have set the OPENID_CLIENT_ID, OPENID_CLIENT_SECRET, OPENID_ISSUER, 
+# OPENID_SESSION_SECRET, and OPENID_SCOPE above
+ALLOW_OPENID=false
+OPENID_LABEL=OpenID
```
Hey as mentioned in the previous PR, you need to have one of the vite enabled prefix's for this to pass through. ie `VITE_` `SCRIPT_` or one of the ones I have setup ie. `ALLOW_` (which is why the other one works).
I'm still not sure if a label is really needed anyway? will people really need to name it something other than OpenID?

bsu3338 replied:
For our organization, we would not name it OpenID.  As a school, we give our credentials a unique name and reference that across various platforms. This way the user does not know the authentication system is LDAP, OpenID, Azure AD, or SAML. It provides consistency across various platforms.  I moved the pull to draft and will make the corrections on the name of the label and retest.

ClaraLeigh commented on client/src/components/Auth/Registration.tsx:
```diff
@@ -1,12 +1,32 @@
+  const [openidLogo, setOpenidLogo] = useState('');
+  const openidLabel = typeof OPENID_LABEL !== 'undefined' ? OPENID_LABEL : 'OpenID';
+
+  useEffect(() => {
+    fetch('/images/openid/logo.svg')
+      .then((response) => {
+        if (response.ok) {
+          return response.url;
+        } else {
+          throw new Error('Logo not found');
+        }
+      })
+      .then((logoUrl) => {
+        setOpenidLogo(logoUrl);
+      })
+      .catch((error) => {
+        console.log('Error:', error);
+      });
+  }, []);
```
can we just include the svg in the project manually? We shouldn't really fetch an image like this

bsu3338 replied:
Could it be a url reference? It could use a default logo if a URL is not given. Again this is connected the label response. By putting our logo on the button it directs our users to naturally use it.  We are working with students and some are less savvy on their computer skills. So anything we can do to make it easier for staff, faculty, and student means less calls to our service desk.

danorlando replied:
btw, I'm finished with the work on that now, so just waiting for it to be reviewed.

bsu3338 commented:
> I ran your branch and got an Internal Server Error on the OAuth callback. Has anyone else been able to successfully test the integration?
> 
> I'll see if I can dig into the root cause. Hopefully user error, but wanted to keep the discussion going on this PR.



> 

Are you using Azure AD?

Does your Web redirect URI have https://fqdn/oauth/openid/callback

bsu3338 committed:
Update .env.example

bsu3338 committed:
Update .env.example

bsu3338 committed:
Update docker-compose.yml

bsu3338 committed:
Update env.d.ts

bsu3338 commented:
I cleaned up the code a little per suggestions from @ClaraLeigh 

I included an SVG from below link for a default logo
https://commons.wikimedia.org/wiki/File:Font_Awesome_5_brands_openid.svg

### Without OpenID label or logo definded
![image](https://github.com/danny-avila/LibreChat/assets/11950213/84d1cdf5-d050-4fc8-adcb-eb9c80b39d42)

### With OpenID label and logo defined
![image](https://github.com/danny-avila/LibreChat/assets/11950213/7f0a3d59-9e5f-4c70-962c-4d7dedfe4c59)

bsu3338 commented on client/src/components/Auth/Login.tsx:
```diff
@@ -1,8 +1,8 @@
 import { useAuthContext } from '~/hooks/AuthContext';
 import { useNavigate } from 'react-router-dom';
-import { SHOW_GOOGLE_LOGIN_OPTION, ALLOW_REGISTRATION, DOMAIN_SERVER } from "~/utils/envConstants";
+import { SHOW_GOOGLE_LOGIN_OPTION, ALLOW_OPENID, VITE_OPENID_LABEL, VITE_OPENID_URL, ALLOW_REGISTRATION, DOMAIN_SERVER } from "~/utils/envConstants";
```
I do not mind updating the code. What is the best way to resync my fork with the current main when there are conflicts?  Last time I did another branch and squashed my changes and resubmitted it.  However, that is probably not the proper method.

danorlando replied:
It should be pretty easy to update once you've synced it.
Steps that I follow:
- I usually just click the sync fork on the main branch from gjthub when I see that its behind
- Then `git checkout main`
- `git pull`
- `git checkout <your_branch>`
- and either `git rebase main` or `git merge main`
I prefer rebase as it makes for a much cleaner commit tree and allows me to step through the commits while resolving conflicts one at a time. It also gives a bit more control in case I want to abort. However, some people find this process difficult if you haven't done it before, in which case you might choose to just merge.

bsu3338 commented:
I converted back to draft for the moment, I am going to resync with main once #518 is approved.  Also, going to look at a possible token issue.

bsu3338 committed:
Update .env.example

bsu3338 committed:
Update .env.example

bsu3338 committed:
Update config.js

bsu3338 committed:
Update Login.tsx

bsu3338 committed:
Update config.js

bsu3338 committed:
Update Login.tsx

bsu3338 committed:
Update Registration.tsx

bsu3338 committed:
Update docker-compose.yml

bsu3338 committed:
Update openidStrategy.js

bsu3338 committed:
Update docker-compose.yml

bsu3338 committed:
Update config.spec.js

bsu3338 committed:
Update Login.spec.tsx

bsu3338 committed:
Update Registration.spec.tsx

bsu3338 committed:
Update types.ts

bsu3338 commented:
@danorlando I think I got everything moved over for #518 

I am opening back up for merging.  I do think I have an issue with the refresh token, but I see a lot of code commented out from a previous commit pertaining to refresh tokens. I do not mind looking into it.  This seems to happen after the user is logged in for a period of time.  They have to logout and login. However, it is really quick.

bsu3338 committed:
Update .env.example

bsu3338 committed:
Update package-lock.json

bsu3338 committed:
Update openidStrategy.js

bsu3338 committed:
Update openidStrategy.js

bsu3338 committed:
Merge branch 'danny-avila:main' into openid-client-sq

bsu3338 committed:
Update config.js

bsu3338 committed:
Update config.js

bsu3338 committed:
Update Login.tsx

bsu3338 committed:
Update Registration.tsx

bsu3338 committed:
Update oauth.js

bsu3338 committed:
Update openidStrategy.js

bsu3338 committed:
Update openidStrategy.js

bsu3338 committed:
Update Registration.tsx

bsu3338 committed:
Update Login.tsx

bsu3338 committed:
Update Login.tsx

bsu3338 committed:
Update Registration.tsx

bsu3338 committed:
Update Registration.tsx

bsu3338 committed:
Merge branch 'danny-avila:main' into openid-client-sq

bsu3338 committed:
Update index.js

bsu3338 committed:
Update index.js

bsu3338 committed:
Update .env.example

bsu3338 commented:
OK, I did a few more checks on the code. I think it functions similar to Facebook and Google. I cleaned up the strategy. I tried removing the session component, but the openid-client errors stating that it requires session support. I do not think the code really uses the session component, because it generates a JWT upon authentication.

bsu3338 committed:
Merge branch 'danny-avila:main' into openid-client-sq

danorlando commented:
> @danorlando I think I got everything moved over for #518
> 
> I am opening back up for merging. I do think I have an issue with the refresh token, but I see a lot of code commented out from a previous commit pertaining to refresh tokens. I do not mind looking into it. This seems to happen after the user is logged in for a period of time. They have to logout and login. However, it is really quick.

Awesome. 

Regarding refresh tokens, I never got as far as fully implementing refresh tokens. As you saw, I had the backend logic for it in place, but the more I worked on the frontend logic, the more I realized that there was more effort required than I had to give for the token management piece. This is something that is really needed for the project though because right now the access token is set to a cookie which is less secure, but what we want is for the refresh token to be set in the cookie and then client exchanges that for an access token. Perhaps this might interest you if you'd like to continue contributing ;)

danny-avila commented:
Regarding package-lock conflict, this is due to me upgrading a dependency. The conflict resolution may not be hard, but in cases it's something crazy like 100+ or even 10+ conflicts, I would delete the package-lock file, delete all node_modules folders in your project directory, and then run npm install to re-generate the package-lock file.

danorlando commented:
> OK, I did a few more checks on the code. I think it functions similar to Facebook and Google. I cleaned up the strategy. I tried removing the session component, but the openid-client errors stating that it requires session support. I do not think the code really uses the session component, because it generates a JWT upon authentication.

yea, I think this is something that differentiates OpenID from standard OAuth2.

danorlando commented:
Any chance you could throw together some setup instructions to [this doc](https://github.com/danny-avila/LibreChat/blob/main/docs/features/user_auth_system.md) similar to  what we have for setting up google auth?

danorlando commented:
> Any chance you could throw together some setup instructions to add to [this doc](https://github.com/danny-avila/LibreChat/blob/main/docs/features/user_auth_system.md) similar to what we have for setting up google auth?

Actually, disregard. Fuegovic is going to add them from the description you created.

fuegovic committed:
Update user_auth_system.md

updated instruction that includes OpenID set up

fuegovic commented:
@bsu3338 @danorlando 
I have submitted the doc, updated with the setup instructions for OpenID.

danorlando commented:
Apologies, I accidentally closed this when I made my last comment.. That was unintentional. Once the final updates ate made and the package-lock conflict is resolved I will approve.

bsu3338 committed:
Update package.json

bsu3338 committed:
Update package-lock.json

fuegovic commented:
> I am opening back up for merging. I do think I have an issue with the refresh token, but I see a lot of code commented out from a previous commit pertaining to refresh tokens. I do not mind looking into it. This seems to happen after the user is logged in for a period of time. They have to logout and login. However, it is really quick.

Please make sure to add the updated documentation to your pull request. I have created a pull request on your fork with the documentation changes, so you can merge them easily.

fuegovic reviewed (changes requested):
Please make sure to add the updated documentation to your pull request. I have created a pull request on your fork with the documentation changes, so you can merge them easily.

the updated doc is also attached here:
[user_auth_system.md](https://github.com/danny-avila/LibreChat/files/11793968/user_auth_system.md)

bsu3338 committed:
Merge pull request #2 from fuegovic/patch-1

Update user_auth_system.md

bsu3338 committed:
Update package-lock.json

bsu3338 committed:
Update package-lock.json

bsu3338 committed:
Update package-lock.json

bsu3338 committed:
Update package-lock.json

bsu3338 committed:
Update package-lock.json

bsu3338 committed:
Update package-lock.json

bsu3338 committed:
Merge remote-tracking branch 'upstream/main' into openid-client-sq

bsu3338 committed:
Update package-lock.json

bsu3338 commented:
I think I have the package-lock.json issue resolved.  I spin up a new docker container each time I build, so I am not fore sure what my original problem was.  I resynced using the main's .json file. Then reupdated with npm install.

bsu3338 commented:
> > > I ran your branch and got an Internal Server Error on the OAuth callback. Has anyone else been able to successfully test the integration?
> > > I'll see if I can dig into the root cause. Hopefully user error, but wanted to keep the discussion going on this PR.
> > 
> > 
> > > 
> > 
> > 
> > Are you using Azure AD?
> > Does your Web redirect URI have https://fqdn/oauth/openid/callback
> 
> Yes, I have that all set up - and tried with latest changes.
> 
> ```dotenv
> OPENID_CLIENT_ID=<my-guid>
> OPENID_CLIENT_SECRET=<my-secret>
> OPENID_ISSUER=https://login.microsoftonline.com/<my-tenant>/v2.0/
> OPENID_SESSION_SECRET=<my-uniquestring>
> OPENID_SCOPE="openid profile email"
> OPENID_CALLBACK_URL=/oauth/openid/callback
> # ...
> DOMAIN_CLIENT=https://<my-shareddns>
> DOMAIN_SERVER=http://localhost:3080
> ```
> 
> I have my callback set in my AAD app registration to be `https://<my-shareddns>/oauth/openid/callback` (and `http://localhost:3080/oauth/openid/callback` for local testing). I have access and ID tokens checked.
> 
> AAD logs me in, and sends me back to http://localhost:3080/oauth/openid/callback and I get an Internal Server Error and get redirected back to localhost (instead of my-shareddns callback):
> 
> ```
> LibreChat         | AxiosError: Request failed with status code 403
> LibreChat         |     at settle (/app/node_modules/axios/dist/node/axios.cjs:1909:12)
> LibreChat         |     at Unzip.handleStreamEnd (/app/node_modules/axios/dist/node/axios.cjs:2989:11)
> LibreChat         |     at Unzip.emit (node:events:525:35)
> LibreChat         |     at endReadableNT (node:internal/streams/readable:1359:12)
> LibreChat         |     at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
> LibreChat         | Error: did not find expected authorization request details in session, req.session["oidc:login.microsoftonline.com"] is undefined
> LibreChat         |     at /app/node_modules/openid-client/lib/passport_strategy.js:132:13
> LibreChat         |     at OpenIDConnectStrategy.authenticate (/app/node_modules/openid-client/lib/passport_strategy.js:191:5)
> LibreChat         |     at attempt (/app/node_modules/passport/lib/middleware/authenticate.js:369:16)
> LibreChat         |     at authenticate (/app/node_modules/passport/lib/middleware/authenticate.js:370:7)
> LibreChat         |     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
> LibreChat         |     at next (/app/node_modules/express/lib/router/route.js:144:13)
> LibreChat         |     at Route.dispatch (/app/node_modules/express/lib/router/route.js:114:3)
> LibreChat         |     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
> LibreChat         |     at /app/node_modules/express/lib/router/index.js:284:15
> LibreChat         |     at Function.process_params (/app/node_modules/express/lib/router/index.js:346:12)
> ```
> 
> If I update the `DOMAIN_SERVER` to match my shared DNS, the OpenID redirect tries to send me to `https://my-shareddns/<my-tenant>/oauth2/v2.0/authorize?client_id=<my-guid>...` which seems ... incorrect (and a 404 error).
> 
> Feels like there's either a few issues, or I'm missing something.

I think I know what it is. It seems to work when a person has a profile picture and error when it does not.  I am looking at the code now to take that into account.

bsu3338 commented:
Converted to draft. Looks like there is an issue if a profile picture is not defined.  I am updating the code to account for that scenario.

bsu3338 committed:
Update openidStrategy.js

bsu3338 commented:
@mitch-b Do you mind trying it now.  I have updated the code to account for a profile picture not being returned.

mitch-b commented:
> @mitch-b Do you mind trying it now. I have updated the code to account for a profile picture not being returned. The axios portion of the OpenID authentication is related to the download of a profile picture. I just noticed that today when one of our accounts was getting a similar error. The updated code fixed it for a local user.

Thanks, @bsu3338. No more error, but I do just get stuck/routed to the `/login` page after the OAuth callback. Am I supposed to register a different way? Should it create a user for me?

bsu3338 commented:
> > @mitch-b Do you mind trying it now. I have updated the code to account for a profile picture not being returned. The axios portion of the OpenID authentication is related to the download of a profile picture. I just noticed that today when one of our accounts was getting a similar error. The updated code fixed it for a local user.
> 
> Thanks, @bsu3338. No more error, but I do just get stuck/routed to the `/login` page after the OAuth callback. Am I supposed to register a different way? Should it create a user for me?

@mitch-b  The user should be auto created. I do not think the "registration page" is really needed if you are going to use only OpenID for authentication.  The user is created the same rather you use the login button on the login page or the one on the registration page.

Try setting you DOMAIN_SERVER back to 
DOMAIN_SERVER=https://<my-shareddns>
to see if it makes a difference now.  

Check the logs at both service startup and during authentication.  Although Microsoft allows localhost in the redirect URI, I am not for sure if the openid-client allows "http".  I will be check sometime tomorrow.

I am running my container behind haproxy for SSL termination, so both my DOMAIN_CLIENT and DOMAIN_SERVER are https://<my-shareddns>

Please paste the error in the reply, and I will try to help troubleshoot more.

mitch-b commented:
Sure, when I set the `DOMAIN_SERVER` to be `https://<my-shareddns>`, when I click the "Login with OpenID" button, it tries to send me to an invalid authority - `https://<my-shareddns>/<my-tenant>/oauth2/v2.0/authorize?client_id=<my-guid>`. Interestingly, when `DOMAIN_SERVER` is `localhost`, it sends me to the valid AAD issuer.

bsu3338 commented:
> Azure AD
> 
> * Sign in to the Azure Portal
> * Search for and select Azure Active Directory
> * Select App registrations > New registration
> * Enter a display Name
> * Choose Web
> * Enter https://fqdn/oauth/openid/callback
> * Select Overview
> * Note of the Application (client) ID and Directory (tenant) ID
> * Select Authentication
> * Check Access tokens and ID tokens
> * Select Certificates & Secrets
> * Create a secret and note the value
> 
> Edit .env to include variables OPENID_CLIENT_ID=From Note Above OPENID_CLIENT_SECRET=From Note Above OPENID_ISSUER=https://login.microsoftonline.com/(tenant id)/v2.0/ OPENID_SESSION_SECRET=Random String OPENID_SCOPE=openid profile email OPENID_CALLBACK_URL=/oauth/openid/callback

@mitch-b I just want to confirm that you did the below steps to create the app

1. Sign in to the Azure Portal
2. Search for and select Azure Active Directory
3. Select App registrations > New registration
4. Enter a display Name
5. Choose Web
6. Enter https://fqdn/oauth/openid/callback
7. Select Overview
8. Note of the Application (client) ID and Directory (tenant) ID
9. Select Authentication
10. Check Access tokens and ID tokens
11. Select Certificates & Secrets
12. Create a secret and note the value

Edit .env to include variables
OPENID_CLIENT_ID=From Note Above
OPENID_CLIENT_SECRET=From Note Above
OPENID_ISSUER=https://login.microsoftonline.com/(tenant id)/v2.0/
OPENID_SESSION_SECRET=Random String
OPENID_SCOPE=openid profile email
OPENID_CALLBACK_URL=/oauth/openid/callback

Also did you grant admin consent for the app?  It should have prompted you on first login.
![image](https://github.com/danny-avila/LibreChat/assets/11950213/ab60ceb4-b16c-44d2-8c2f-f250347aa77c)

bsu3338 committed:
Merge branch 'danny-avila:main' into openid-client-sq

mitch-b commented:
> @mitch-b I just want to confirm that you did the below steps to create the app
> 
> 1. Sign in to the Azure Portal
> 2. Search for and select Azure Active Directory
> 3. Select App registrations > New registration
> 4. Enter a display Name
> 5. Choose Web
> 6. Enter https://fqdn/oauth/openid/callback
> 7. Select Overview
> 8. Note of the Application (client) ID and Directory (tenant) ID
> 9. Select Authentication
> 10. Check Access tokens and ID tokens
> 11. Select Certificates & Secrets
> 12. Create a secret and note the value
> 
> Edit .env to include variables OPENID_CLIENT_ID=From Note Above OPENID_CLIENT_SECRET=From Note Above OPENID_ISSUER=https://login.microsoftonline.com/(tenant id)/v2.0/ OPENID_SESSION_SECRET=Random String OPENID_SCOPE=openid profile email OPENID_CALLBACK_URL=/oauth/openid/callback
> 
> Also did you grant admin consent for the app? It should have prompted you on first login. ![image](https://user-images.githubusercontent.com/11950213/247611281-ab60ceb4-b16c-44d2-8c2f-f250347aa77c.png)
> 
> Does your user have an email account associated with it?

Yes, that all is good. What I'm experiencing is within LibreChat itself... passport is sending me to my own DNS name instead of login.microsoftonline.com when my `DOMAIN_SERVER` is anything but localhost. That doesn't feel like an AAD registration setting to me.

bsu3338 committed:
Update openidStrategy.js

Lookup user based on openID instead of email.  This is because not all AzureAD users may have an email tied to their account

bsu3338 commented:
@mitch-b 

When you go to 
https://login.microsoftonline.com/\<tenant-id\>/v2.0/.well-known/openid-configuration

Do you see 
issuer: https://login.microsoftonline.com/\<tenant-id\>/v2.0
as an option

mitch-b commented:
Yes, I do see that issuer there. 

![image](https://github.com/danny-avila/LibreChat/assets/1643072/95172400-2f73-44bc-b60b-8d7360fce1b8)

Let me grab the behavior I see. Given my `.env` shows:

```env
DOMAIN_SERVER=https://chatgpt.randomdomain.com
```

I can navigate to that address, and if I hover the OpenID button, I see `https://chatgpt.randomdomain.com/oauth/openid`:

![image](https://github.com/danny-avila/LibreChat/assets/1643072/70c1312e-afc9-4e9d-a89b-47d9bbb28d7e)

You can see that the server sends me a 302 to my own `DOMAIN_SERVER` with all the important information that should be going to my AAD Issuer ... not LibreChat itself. 

![image](https://github.com/danny-avila/LibreChat/assets/1643072/9c1fcc71-018e-4ef5-8cf7-3f67e1799431)

But interestingly - if I put my `DOMAIN_SERVER` as `http://localhost:3080`, it seems to send me to AAD to sign in!

bsu3338 committed:
Update openidStrategy.js

First try to match an email, then try openIdID

bsu3338 committed:
Update openidStrategy.js

bsu3338 committed:
Update openidStrategy.js

Consider a family name or given name is not provided

bsu3338 commented:
@mitch-b 

I see what you mean, but I am at a little bit of a loss. I setup a second server. So now I have a test and production version. The code is working on both.   I have included a image similar to yours for comparison.



![image](https://github.com/danny-avila/LibreChat/assets/11950213/eb283e4d-287e-4729-ae93-47135b74b895)

![image](https://github.com/danny-avila/LibreChat/assets/11950213/53a0912f-2eb0-4d4c-a7c7-460e6ffe947f)

bsu3338 committed:
Merge branch 'danny-avila:main' into openid-client-sq

bsu3338 committed:
Merge branch 'danny-avila:main' into openid-client-sq

mitch-b commented:
Thanks, @bsu3338 - I wonder if it's something on my side, then. I'm investigating.

mitch-b commented:
So even though I'm temporarily routed to my own DNS for the login - it could be a proxy issue. When I manually swap my custom DNS for `login.microsoftonline.com`, I am granted a token, and redirected properly and my user is created! 

So good news on the functionality front, figuring out my routing issue right now.

danorlando commented:
Is this ready to be merged? If not, what is it waiting on?

fuegovic commented:
@danorlando afaik it's waiting for another confirmation of success before it can be merged

danny-avila commented:
I got it working and tested all other logins, including regular sign up, everything works as expected

My only feedback is that this should not include the additional `--- OR ----` but this can be fixed in another PR
![image](https://github.com/danny-avila/LibreChat/assets/110412045/5729bf4e-afb5-4cb9-b1f6-950344e4635c)

also the doc should be updated to make the oauth callback localhost:3080 as it can be confusing with `fqdn` otherwise, also for another PR

I did experience an Internal Server Error on first login, maybe this was the issue? 
```bash 

Timestamp: 2023-06-25 02:38:11Z)
    at processResponse (C:\Users\Danny\Documents\chatgpt-clone\node_modules\openid-client\lib\helpers\process_response.js:38:13)
    at Client.grant (C:\Users\Danny\Documents\chatgpt-clone\node_modules\openid-client\lib\client.js:1327:22)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Client.callback (C:\Users\Danny\Documents\chatgpt-clone\node_modules\openid-client\lib\client.js:476:24)     
    at async C:\Users\Danny\Documents\chatgpt-clone\node_modules\openid-client\lib\passport_strategy.js:166:22
Error: did not find expected authorization request details in session, req.session["oidc:login.microsoftonline.com"] is undefined
    at C:\Users\Danny\Documents\chatgpt-clone\node_modules\openid-client\lib\passport_strategy.js:132:13
    at OpenIDConnectStrategy.authenticate (C:\Users\Danny\Documents\chatgpt-clone\node_modules\openid-client\lib\passport_strategy.js:191:5)
    at attempt (C:\Users\Danny\Documents\chatgpt-clone\node_modules\passport\lib\middleware\authenticate.js:369:16)       
    at authenticate (C:\Users\Danny\Documents\chatgpt-clone\node_modules\passport\lib\middleware\authenticate.js:370:7)   
    at Layer.handle [as handle_request] (C:\Users\Danny\Documents\chatgpt-clone\node_modules\express\lib\router\layer.js:95:5)
    at next (C:\Users\Danny\Documents\chatgpt-clone\node_modules\express\lib\router\route.js:144:13)
    at Route.dispatch (C:\Users\Danny\Documents\chatgpt-clone\node_modules\express\lib\router\route.js:114:3)
    at Layer.handle [as handle_request] (C:\Users\Danny\Documents\chatgpt-clone\node_modules\express\lib\router\layer.js:95:5)
    at C:\Users\Danny\Documents\chatgpt-clone\node_modules\express\lib\router\index.js:284:15
    at Function.process_params (C:\Users\Danny\Documents\chatgpt-clone\node_modules\express\lib\router\index.js:346:12)   
Error downloading image at URL "https://graph.microsoft.com/v1.0/me/photo/$value": AxiosError: Request failed with status code 403
```

on second attempt, i could login fine so I'm just going to go ahead and merge.

Thank you very much, excellent work and thanks for being patient on this merging

bsu3338 commented:
I am glad it did take a little bit, the code improved as @mitch-b tested it and the suggestions from all the existing developers where helpful.
